### PR TITLE
[FLINK-1422] Add withParameters() to documentation

### PR DESCRIPTION
**Pull request by zentol**

### CODE REVIEW

**Comment by aljoscha** [comment]
+1 looks good to me

**Comment by fhueske** [comment]
Very good addition to the documentation! 
I also agree with Alex (see FLINK-1422) to describe how functions can be configured via the constructor, i.e., that function objects are serialized and distributed. This is actually much more comfortable and should therefore be the preferred way. The Configuration object is rather a relict of former Stratosphere times ;-)

Another small improvement would be to put a link to the Function configuration section into the "Passing Function to Flink" section.

**Comment by zentol** [comment]
gotcha, I'll address the points mentioned.

**Comment by zentol** [comment]
updated.

File: docs/programming_guide.md [hunk 1]
```diff
@@ -2398,6 +2399,61 @@ of a function, or use the `withParameters(...)` method to pass in a configuratio
 [Back to top](#top)
 
 
+Passing parameters to functions
+-------------------
+
+Parameters can be passed to rich functions using either the constructor (if the function is
```
Comment (fhueske) [inline docs/programming_guide.md]: Only the Configuration object method requires a RichFunction to overwrite `open()`. Configuring via the constructor (or any other method which sets non-transient fields) works also for functions that implement the function interfaces. Also it is not required that the class is defined as `static final`. That is only necessary if the class is defined inside another class.

Comment (zentol) [inline docs/programming_guide.md]: ok, so which requirements do exist for the operator to be serializable? or can i just omit those details and say something like "parameters can be stored in non-transient fields if the function is serializable".

Comment (fhueske) [inline docs/programming_guide.md]: If the function is not serialisable the program will fail in any case. That is not related to the configuration. 
IMO, the documentation should state that the function object is serialised using standard java serialization and shipped "as-is" to all parallel task instances. Mentioning that transient fields are not shipped is OK but not mandatory (somebody who uses the transient keyword should know what it does).

**Comment by uce** [comment]
Thanks, Chesnay! I've added the constructor parametrization, fixed the code examples, and rebased on the lastest master.

Going to merge it.

**Pull request closed without merging**